src/sys/ioctl/mod.rs

```diff
@@ -1,102 +1,175 @@
-//!     pub speed_hz: u32,
-//!     pub delay_usecs: u16,
-//!     pub bits_per_word: u8,
-//!     pub cs_change: u8,
-//!     pub pad: u32,
-//! }
+//! # #[macro_use] extern crate nix;
+//! # const SPI_IOC_MAGIC: nix::libc::c_int = 'k' as nix::libc::c_int;
+//! ioctl!(read spi_read_mode with SPI_IOC_MAGIC, 1; u8);
+//! # fn main() {}
+//! ```
+//!
+//! This generates a function like:
 //!
-//! #[cfg(linux)]
-//! mod ioctl {
-//!     use super::*;
-//!
-//!     const SPI_IOC_MAGIC: u8 = 'k' as u8;
-//!     const SPI_IOC_NR_TRANSFER: u8 = 0;
-//!     const SPI_IOC_NR_MODE: u8 = 1;
-//!     const SPI_IOC_NR_LSB_FIRST: u8 = 2;
-//!     const SPI_IOC_NR_BITS_PER_WORD: u8 = 3;
-//!     const SPI_IOC_NR_MAX_SPEED_HZ: u8 = 4;
-//!     const SPI_IOC_NR_MODE32: u8 = 5;
-//!
-//!     ioctl!(read  get_mode_u8 with SPI_IOC_MAGIC, SPI_IOC_NR_MODE; u8);
-//!     ioctl!(read  get_mode_u32 with SPI_IOC_MAGIC, SPI_IOC_NR_MODE; u32);
-//!     ioctl!(write set_mode_u8 with SPI_IOC_MAGIC, SPI_IOC_NR_MODE; u8);
-//!     ioctl!(write set_mode_u32 with SPI_IOC_MAGIC, SPI_IOC_NR_MODE32; u32);
-//!     ioctl!(read  get_lsb_first with SPI_IOC_MAGIC, SPI_IOC_NR_LSB_FIRST; u8);
-//!     ioctl!(write set_lsb_first with SPI_IOC_MAGIC, SPI_IOC_NR_LSB_FIRST; u8);
-//!     ioctl!(read  get_bits_per_word with SPI_IOC_MAGIC, SPI_IOC_NR_BITS_PER_WORD; u8);
-//!     ioctl!(write set_bits_per_word with SPI_IOC_MAGIC, SPI_IOC_NR_BITS_PER_WORD; u8);
-//!     ioctl!(read  get_max_speed_hz with SPI_IOC_MAGIC, SPI_IOC_NR_MAX_SPEED_HZ; u32);
-//!     ioctl!(write set_max_speed_hz with SPI_IOC_MAGIC, SPI_IOC_NR_MAX_SPEED_HZ; u32);
-//!     ioctl!(write spidev_transfer with SPI_IOC_MAGIC, SPI_IOC_NR_TRANSFER; spi_ioc_transfer);
-//!     ioctl!(write buf spidev_transfer_buf with SPI_IOC_MAGIC, SPI_IOC_NR_TRANSFER; spi_ioc_transfer);
+//! ```
+//! # #[macro_use] extern crate nix;
+//! # use nix::libc::c_int as c_int;
+//! # use nix::libc::c_ulong as c_ulong;
+//! # use std::mem;
+//! # use nix::{Errno, libc, Result};
+//! # const SPI_IOC_MAGIC: u8 = 'k' as u8;
+//! pub unsafe fn spi_read_mode(fd: c_int, data: *mut u8) -> Result<c_int> {
+//!     let res = libc::ioctl(fd, ior!(SPI_IOC_MAGIC, 1, mem::size_of::<u8>()), data);
+//!     Errno::result(res)
 //! }
+//! # fn main() {}
+//! ```
+//!
+//! The return value for `ioctl` functions generated by the `ioctl!` macro are assumed to return -1
+//! on error and everything else is a valid return value. If the return value needs to be changed,
+//! you can use `Result::map` on the return value in a helper function.
 //!
-//! // doctest workaround
-//! fn main() {}
+//! There are equivalent forms for `write`, `none` (no data in or out), and `readwrite`. The mode
+//! for a given `ioctl` should be clear from the documentation if it has good documentation.
+//! Otherwise it will be clear based on the macro used to generate the `ioctl` number where
+//! `_IOC`, _IOR`, `_IOW`, and `_IORW` map to "none", "read", "write", and "readwrite"
```

Comment:
`_IO` maps to "none", not `_IOC`. `_IOC` is a macro that puts all the components together to create a single IOCTL number.